### PR TITLE
Alerting: edit cloud receivers

### PR DESCRIPTION
**Pull request by domasx2**
Editing of alertmanagers receivers!

Sorry about the size. I intended this to be smaller, but by the time all problems were solved and tests written... 

Achieved by:
* extending our notification "form builder" with fields required for editing alertmanager notifier forms: subform, subform array, key:value map, string array
* hardcoding prom alertmanager notifier configs in `cloud-alertmanager-notifier-types`

Bonus:
* Improvements to display backend error messages when there is an array of them

Known issues:
*  Cortex alertmanager secrets are returned redacted to `"secret"`. Currently there is no way to *not* submit them and have backend pick up previously set secrets. Thus on every update all secrets have to be re-entered... Working on it
* Backend translates json to yaml and back when proxying requests. Seems some structs are translated incorrectly and data is lost. For example `http_config` , `slack_configs.0.actions.confirm` 
* ~~If there are multiple receivers, deleting any but the last receiver messes up the form. Working on this. `react-hook-form` is poop.~~ react-hook-form protip: soft delete items, but keep them registered and filter out on submit
* ~~Doesn't work if there is no user config uploaded in the first place (fresh alertmanager instance). Will be addressed in a separate PR, waiting for a backend endpoint to be merged~~ made it fill in some defaults

Todos:
- [x] add a test. This form is so brittle that feels some tests are must have
- [x] fix the deleting receiver problem. It seems that `react-hook-form` cannot handle deleting objects from array in case these objects have differing fields. Speculating, race condition between field register/unregister which are triggerd on render/unmount. Considering internally marking items as "deleted" without actually removing them form the array, then filtering them out on submit.

https://user-images.githubusercontent.com/847684/116884656-a6fd7180-ac2f-11eb-9f94-9bc8ffaaf178.mp4

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/alerting/unified/components/receivers/TemplatesTable.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ export const TemplatesTable: FC<Props> = ({ config, alertManagerName }) => {
                   <td>{name}</td>
                   <td className={tableStyles.actionsCell}>
                     <ActionIcon
```
Comment (torkelo) [inline public/app/features/alerting/unified/components/receivers/TemplatesTable.tsx:61]: ActionIcon? We have a ButtonIcon component, is this similar?

**Review by torkelo — COMMENTED** [review]
nice tests!

File: public/app/features/alerting/unified/components/receivers/TemplatesTable.tsx (hunk 1 above)
Comment (domasx2) [reply public/app/features/alerting/unified/components/receivers/TemplatesTable.tsx:61]: It is an `<Icon />` wrapped in `<Tooltip/>` and optionally wrapped in `<Link />`, I did not find a component to do both of these

**Review by mckn — COMMENTED** [review]
Not sure how to approach this review. When I start to test this it feels like we first of all need to improve the UX because in the current form you basically can end up with a huge form with lots of fields on the same page.

Why is this a problem? It get way to complex for a regular user to fill in the form. An example: add a couple of contact points, scroll down and try to save it. If the user have entered anything that results in an error that message is probably displayed above the fold and need to scroll up to see what the error is.

Some ideas on how to improve this:

- Simplify the form so you only can create one contact point per page and introduce a contact point group if you want to group multiple of them together.

- Move things like SMPT etc. that you probably only want to enter once to a settings page.

- Introduce an advance section for things that are optional and only used in rare cases.

With this said, I'm not sure how we should proceed?

File: public/app/features/alerting/unified/components/receivers/ReceiversTable.tsx [hunk 2]
```diff
@@ -34,7 +34,7 @@ export const ReceiversTable: FC<Props> = ({ config, alertManagerName }) => {
       addButtonLabel="New contact point"
```
Comment (mckn) [inline public/app/features/alerting/unified/components/receivers/ReceiversTable.tsx:34]: I'm a bit confused by the lingio in the code. Looks like this functionality is called "ContactPoint" in the UI but we call it Receivers in the code. How come?

**Comment by domasx2** [comment]
> Not sure how to approach this review. When I start to test this it feels like we first of all need to improve the UX because in the current form you basically can end up with a huge form with lots of fields on the same page.

Most fields are collapsed by default, I don't think it's too bad. Also there is just no time to go back to the drawing board at this stage


> Why is this a problem? It get way to complex for a regular user to fill in the form. An example: add a couple of contact points, scroll down and try to save it. If the user have entered anything that results in an error that message is probably displayed above the fold and need to scroll up to see what the error is.

It jumps to the first error automatically

> * Simplify the form so you only can create one contact point per page and introduce a contact point group if you want to group multiple of them together.

I think having multiple contact point types per contact point is rare-ish, I wouldn't worry about it now

> * Move things like SMPT etc. that you probably only want to enter once to a settings page.

Thats pretty much how it will be, there will be [separate form for global settings](https://github.com/grafana/alerting-squad/issues/113). You'll be able to override them for each contact point type individually though.

> * Introduce an advance section for things that are optional and only used in rare cases.

That's how it works already, most settings are collapsed by default in a section
image


> With this said, I'm not sure how we should proceed?

I would say check for critical problems and approve. There's no time for any major changes at the moment. We'll reasses and improve later

**Comment by mckn** [comment]
@domasx2 ok ok, super! I will continue the review and we can rework this after the v8 release. I'm just starting to get worried about that the "rework" will need to be done in the backend APIs as well.

**Comment by domasx2** [comment]
> @domasx2 ok ok, super! I will continue the review and we can rework this after the v8 release. I'm just starting to get worried about that the "rework" will need to be done in the backend APIs as well.

Grafana is aligning with cortex/prometheus API, so I don't think there's much danger of radical change there :) and it does kinda limit what our options on the frontend are

File: public/app/features/alerting/unified/utils/receiver-form.ts [hunk 3]
```diff
@@ -32,6 +37,40 @@ export function grafanaReceiverToFormValues(
   return [values, channelMap];
 }
 
+export function cloudReceiverToFormValues(
```
Comment (mckn) [inline public/app/features/alerting/unified/utils/receiver-form.ts:40]: nit: maybe give this function some <3 to make it easier to read.

**Review by mckn — APPROVED** [review]
Really great work with this @domasx2! You are a beast when it comes to getting features out the door. Left one small nit but overall great work. Love the tests!

File: public/app/features/alerting/unified/utils/receiver-form.ts (hunk 3 above)
Comment (domasx2) [reply public/app/features/alerting/unified/utils/receiver-form.ts:40]: I added more comments for each step of the transform. Could not figure out how to break it down some more meaningfully though :/

**Pull request merged**